lib/src/bindings.dart

```diff
@@ -217,3 +225,36 @@ abstract class DeltaDocumentSnapshot implements admin.DocumentSnapshot {
   /// Gets the value for a given key.
   external dynamic get(key);
 }
+
+@JS()
+@anonymous
+abstract class PubsubFunctions {
+  /// Registers a function that triggers on Pubsub write events to
+  /// the [topic].
+  external TopicBuilder topic(String topic);
+}
+
+/// The Pubsub topic builder interface.
+@JS()
+@anonymous
+abstract class TopicBuilder {
+  /// Event handler that fires every time an event is public in Pubsub.
+  external CloudFunction onPublish(void handler(Message message));
+}
+
+/// Interface representing a Google Cloud Pub/Sub message.
+@JS()
+@anonymous
+abstract class Message {
+  /// User-defined attributes published with the message, if any.
+  external Map<String, String> get attributes;
```

Comment:
I don't think JS interop supports `Map`. According to docs this field should contain plain JS object and I usually use `dynamic` in this case (unless it's an object with explicitly defined set of fields, e.g. some sort of options object or a data class). In this case attributes can be whatever so I'd just put `dynamic`.

On the Dart side (the other `Message` class which wraps this one) this field needs to be converted into a `Map<String, dynamic>` using `dartify`.

Reply:
From what I can tell, attributes is always `Map<String, String>`, so I think it makes sense to leave it as such. What do you think?

Reply:
Would definitely be fine with me if Dart2 / dartdevc are gonna be happy as well.

Some background: `dartify` creates an instance of `Map<String, dynamic>` because that's as far as it can go in terms fo types. However `Map<String, dynamic>` is not assignable to `Map<String, String>` according to the new type rules in Dart2.

I'd try to test it and see if there are type check warnings and if there are we'd just need to cast it with something like:

```dart
Map<String, String> get attributes => 
  new Map<String, String>.from(dartify(nativeInstance.attributes));
```

Reply:
BTW, not sure if you tried compiling with DDC. It did compile for me, but deploying with firebase tools failed complaining it couldn't find dart_sdk module. It's definitely not a big issue as DDC output is not supposed to be used on production, but I've been using it at least to check if there are any type errors during compilation.